### PR TITLE
Solve multiple timing-related Eject Pack bugs

### PR DESCRIPTION
**Pull request by singiamtel**
As suggested by: [Various Eject Pack bugs](https://github.com/smogon/pokemon-showdown/projects/3#card-70964079)

Thanks to the amazing DaWoblefet [End-turn resolution order research](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/page-64#post-9244179)

### CODE REVIEW

**Comment by singiamtel** [comment]
Same as the other PR, I didn't break the tests, I swear D:

**Comment by singiamtel** [comment]
image

It shouldn't, it looks fine to me

**Comment by MathyFurret** [comment]
Tested this locally and it does have the issue pyuk pointed out

image

**Comment by singiamtel** [comment]
Oh, I was testing on the wrong branch locally...
Embarrasing, my bad. Working on fixing it

**Comment by DaWoblefet** [comment]
> Same as the other PR, I didn't break the tests, I swear D:

Looks to be a Typescript update; we'll get it resolved soon.

File: test/sim/items/ejectpack.js [hunk 1]
```diff
@@ -15,10 +15,13 @@ describe(`Eject Pack`, function () {
 			{species: 'Magikarp', item: 'ejectpack', moves: ['splash']},
 			{species: 'Mew', moves: ['splash']},
 		], [
-			{species: 'Machop', moves: ['leer']},
+			{species: 'Regieleki', moves: ['leer']},
 		]]);
 		battle.makeChoices();
+		const magikarp = battle.p1.active[0];
 		assert.equal(battle.p1.requestState, 'switch');
```
Comment (singiamtel) [inline test/sim/items/ejectpack.js:22]: I expanded the test a bit to check if the switch was done at the right time

File: data/items.ts [hunk 2]
```diff
@@ -1420,20 +1420,43 @@ export const Items: {[itemid: string]: ItemData} = {
...
+					target.abilityState.eject = true;
 				}
 			}
-			if (eject) {
+			if (target.abilityState.eject) {
 				if (target.hp) {
 					if (!this.canSwitch(target.side)) return;
-					for (const pokemon of this.getAllActive()) {
-						if (pokemon.switchFlag === true) return;
+					let lastToMove = true;
```
Comment (singiamtel) [inline data/items.ts:1433]: Not sure if this is the best way to do it. My approach is not switching in OnAfterBoost if you were the last one to move, in order to allow some end of turn effects like Power Construct to happen before the switch. I'm using this check to switch on OnResidual if that's the case

**Pull request closed without merging**